### PR TITLE
Introducing a new Delimiter

### PR DESCRIPTION
**Pull request by schlunz**
Dear @otoolep,

as discussed in #25, I was planning to improve the delimiter by using a new method. This one is basically a netstrings implementation, which means that all log messages must be send like this:

`{number of bytes}` + `:` + `{bytes}` + `;`

Luckily, there are libraries in each programming language, which encode the strings on the client site.
The new delimiter brings the following advantages:

- it can separate log messages from each other in a very save way
- it is easy to implement the decoding part
- it does not need to match regexp
- it breaks, when decoding errors happen (will be interesting in the next part of the description)

So lets talk about *backwards compatibility*.

Since the new delimiter breaks, when an decoding errors happen, the `input/collector.go` switches to the old delimiter (now labeled as `fallback_delimiter`), which means that:

- if a client sends logs the old fashion way, the *delimiter* will break instantly and the *fallback* will handle all following logs
- if a client sends logs netstrings like encoded, but for some stupid reason (the encoded bytes contain encoding errors), the *delimiter* will break instantly, when such error happensand the *fallback* will handle all following logs
- if everything is encoded fine, logs will be processed a little bit faster (since regexp matching is not needed anymore).

I think this change is important to make the input process faster and  to make sure there is a way to delimit json.

**Note:** this PR requires #30 and only implements the new delimiter (there is no part in the code were its actually used).

I hope you like the changes,

sincerely schlunz.

### CODE REVIEW

File: input/delimiter/delimiter.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	mutex      *sync.Mutex = &sync.Mutex{}
+	brokenErr              = errors.New("broken")
+	lenIncErr              = errors.New("length-buffer-incomplete")
+	lenInvErr              = errors.New("length-buffer-invalid-byte")
+	lenConvErr             = errors.New("length-buffer-conversion-error")
+	valIncErr              = errors.New("value-buffer-incomplete")
+)
+
+// A Delimiter detects when message lines start.
+type Delimiter struct {
```
Comment (otoolep) [inline input/delimiter/delimiter.go]: This needs a better name than just Delimiter. It's following a particular scheme, i.e. "netstrings".

**Comment by otoolep** [comment]
I'm not convinced that this form of syslog-line delimiter is important enough to support it automatically. Are there any log systems out that generate this?

I'm fine with adding it, but it should be enabled by a switch on the command line.

**Comment by schlunz** [comment]
@otoolep I renamed the "netstrings like delimiter" to be the `NetstrDelimiter{}`. The next PR will provide integration for this delimiter, as well as a flag to enable it.

I don't know any log system that uses such delimiter method, but there are some libraries, such as [msgpack/msgpack](https://github.com/msgpack/msgpack/blob/master/spec.md#formats-str) or the [chunked transfer encoding](https://en.wikipedia.org/wiki/Chunked_transfer_encoding#Format) in HTTP1.1, which take use of similar methods.

File: input/netstr_delimiter.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+// NewNetstrDelimiter returns an initialized NetstrDelimiter.
+func NewNetstrDelimiter() *NetstrDelimiter {
+	return &NetstrDelimiter{
+		lenBuff: *bytes.NewBuffer([]byte{}),
+		valBuff: *bytes.NewBuffer([]byte{}),
+	}
+}
+
+// Returns rather a new result is available
```
Comment (otoolep) [inline input/netstr_delimiter.go]: GoDoc: the first word in this comment should be "Push".

File: input/netstr_delimiter.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+// and the first occurring error (if any occurred).
+func (d *NetstrDelimiter) Push(b byte) (bool, error) {
+	if d.brokenMode {
+		return NoResult, errors.New("broken")
+	}
+	return d.processByte(b)
+}
+
+// Resets the instance close to its initial state.
```
Comment (otoolep) [inline input/netstr_delimiter.go]: Resets -> Reset, as per GoDoc.

**Comment by otoolep** [comment]
Thanks.

The GoDoc comments are not quite correct. They should start with the name of the function, followed by the description. Can you fix that up, and also rebase on top of master?

**Comment by schlunz** [comment]
@otoolep, thank You for reviewing. I updated the comments.

**Comment by otoolep** [comment]
Great, thanks @schlunz

**Pull request merged**